# -*- coding: utf-8 -*-
"""Visualization.ipynb

Automatically generated by Colab.

Original file is located at
    https://colab.research.google.com/drive/1CBVu9BAspOI75lLwu6TZpQaR_iy2GivH

Use the normalized gene expression dataset to plot a clustered heatmap of the top differentially expressed genes between HBR and UHR samples.
"""

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
gene_exp=pd.read_csv("Data_set/hbr_uhr_top_deg_normalized_counts.csv")
gene_exp.rename(columns={"Unnamed: 0":"genes"},inplace=True)
gene_exp.set_index("genes",inplace=True)
sns.clustermap(gene_exp,cmap="Blues")

"""b. Volcano Plot
Plot log2FoldChange vs log10(Padj) from the DEG results.
Color points by significance:
Upregulated: green
Downregulated: orange
Not significant: grey
Add dashed vertical lines at log2FoldChange = Â±1.
"""

lfc=pd.read_csv("Data_set/hbr_uhr_deg_chr22_with_significance.csv")
#print(lfc.loc[0:5,"significance"])
color_pal={"down":"orange", "ns":"grey","up":"green"}
sns.scatterplot(data=lfc,x="log2FoldChange",y="-log10PAdj",hue="significance",palette=color_pal)
plt.axvline(x=1,linestyle='--')
plt.axvline(x=-1,linestyle='--')

"""c. Scatter Plot (radius vs texture)
Plot texture_mean vs radius_mean and color points by diagnosis (M = malignant, B = benign).
e. Scatter Plot (smoothness vs compactness)
Plot compactness_mean vs smoothness_mean colored by diagnosis.
Include gridlines and clear axis labels.
"""

breast_cancer=pd.read_csv("Data_set/data-3.csv")
print(breast_cancer.columns)
sns.scatterplot(data=breast_cancer,y="texture_mean",x='radius_mean',hue='diagnosis')
sns.set_style("whitegrid")
sns.scatterplot(data=breast_cancer,x='smoothness_mean',y='compactness_mean',hue='diagnosis')

"""f. Density Plot (area distribution)
Plot kernel density estimates (KDE) of area_mean for both M and B diagnoses on the same axis.
Add legend and labeled axes.
"""

sns.kdeplot(data=breast_cancer,x='area_mean',hue='diagnosis',fill=True)

"""d. Correlation Heatmap
Compute the correlation matrix of six key features:

radius_mean, texture_mean, perimeter_mean, area_mean, smoothness_mean, compactness_mean.

Plot as a heatmap with correlation values annotated.
"""

breastcancer_sub=breast_cancer[["radius_mean", "texture_mean", "perimeter_mean", "area_mean", "smoothness_mean", "compactness_mean"]]
breastcancer_sub_corr=breastcancer_sub.corr()
sns.heatmap(breastcancer_sub_corr,annot=True,cmap="Blues")